Work with records (#2982)
influxdata/docs-v2: content/flux/v0.x/data-types/composite/record.md

    parent: Composite types
weight: 201
flux/v0.x/tags: ["composite types", "data types"]
---

> A **record** type is a set of key-value pairs (also known as properties).

Reviewer: Should we define properties in our glossary (are we using this term throughout the docs in the same way?). 
Noticed our definition of field refers to "the key-value pair in InfluxDB’s data structure that records metadata and the actual data value." 

Would it be accurate to say a record is a set of fields (or tags?) also known as properties? If not, maybe helpful to clarify this distinction as appropriate throughout.

Author: No, the term "properties" is specific to records. The terms "fields" or "tags" don't really have meaning within the Flux language. They are concepts tied to the InfluxDB storage engine.

Keys (also known as labels) are strings.
Values can be any data type.
Each property can have a different value type.

###### On this page:
- [Record syntax](#record-syntax)
- [Reference values in a record](#reference-values-in-a-record)
- [Operate on records](#operate-on-records)

## Record syntax
A **record** literal contains a set of key-value pairs (properties) enclosed in curly brackets (`{}`).
Properties are comma-delimitted.
**Property keys must be strings** and can optionally be enclosed in double quotes (`"`).

> If a property key contains only numeric characters or any whitespace characters,
> it must be enclosed in double quotes.

Reviewer: ```suggestion
If a property key contains only numeric characters and whitespace characters,
you must enclose the property key in double quotes.
```

Property keys are associated to values by a colon (`:`).
**Values can be any type**.

##### Example records
```js
{foo: "bar", baz: 123.4, quz: -2}

{"Company Name": "ACME", "Street Address": "123 Main St.", id: 1123445}
```

## Reference values in a record
Flux records are **key-indexed**.
To reference values in a record, use **dot notation** or **bracket notation**
and specify the key to reference.

- [Dot notation](#dot-notation)
- [Bracket notation](#bracket-notation)
- [Reference nested records](#reference-nested-records)
- [Reference keys statically](#reference-keys-statically)

### Dot notation
Specify the record to access followed by a period (`.`) and the property key.

```js
c = {
  name: "John Doe",
  address: "123 Main St.",
  id: 1123445
}

c.name
// Returns John Doe

c.id
// Returns 1123445
```

### Bracket notation
Specify the record to access followed by the property key enclosed in double
quotes and square brackets (`[""]`).

{{% note %}}
Use bracket notation to access keys with special or whitespace characters.
{{% /note %}}

```js
c = {
  "Company Name": "ACME",
  "Street Address": "123 Main St.",
  id: 1123445
}

c["Company Name"]
// Returns ACME

c["id"]

> // Returns 1123445

Reviewer: did you mean to use Street Address as and example....or was this intentional to show you can get the id with bracket or dot notation?

Author: Both :). It was to show how to format a key that has a space it it. You have to wrap it in quotes.
